op-deployer/pkg/deployer/state/intent.go

```diff
@@ -32,35 +33,313 @@ func (d DeploymentStrategy) Check() error {
+		}
+		if c.Chains[chainIndex].Roles.L1ProxyAdminOwner != l1ProxyAdminOwner {
+			return fmt.Errorf("invalid l1ProxyAdminOwner address for chain: %s", c.Chains[chainIndex].ID)
+		}
+	}
+
+	return nil
+}
+
+// Ensures the following:
+//  1. no zero-values for non-standard fields (user should have populated these)
+//  2. no non-standard values for standard fields (user should not have changed these)
+func (c *Intent) validateStandardValues() error {
+	standardSuperchainRoles, err := getStandardSuperchainRoles(c.L1ChainID)
+	if err != nil {
+		return fmt.Errorf("error getting standard superchain roles: %w", err)
+	}
+	if *c.SuperchainRoles != *standardSuperchainRoles {
+		return fmt.Errorf("SuperchainRoles does not match standard value")
+	}
+
+	for _, chain := range c.Chains {
+		if chain.ID == emptyHash {
+			return fmt.Errorf("missing l2 chain ID")
+		}
+		if err := chain.Roles.CheckNoZeroAddresses(); err != nil {
+			return err
+		}
+		if chain.Eip1559DenominatorCanyon != standard.Eip1559DenominatorCanyon ||
+			chain.Eip1559Denominator != standard.Eip1559Denominator ||
+			chain.Eip1559Elasticity != standard.Eip1559Elasticity {
+			return fmt.Errorf("%w: chainId=%s", ErrNonStandardValue, chain.ID)
+		}
+		if chain.BaseFeeVaultRecipient == emptyAddress ||
+			chain.L1FeeVaultRecipient == emptyAddress ||
+			chain.SequencerFeeVaultRecipient == emptyAddress {
+			return fmt.Errorf("%w: chainId=%s", ErrFeeVaultZeroAddress, chain.ID)
+		}
+	}
+
+	return nil
+}
+
+func getStandardSuperchainRoles(l1ChainId uint64) (*SuperchainRoles, error) {
+	superCfg, err := standard.SuperchainFor(l1ChainId)
+	if err != nil {
+		return nil, fmt.Errorf("error getting superchain config: %w", err)
+	}
+
+	proxyAdmin, _ := standard.ManagerOwnerAddrFor(l1ChainId)
+	guardian, _ := standard.GuardianAddressFor(l1ChainId)
+
+	superchainRoles := &SuperchainRoles{
+		ProxyAdminOwner:       proxyAdmin,
+		ProtocolVersionsOwner: common.Address(*superCfg.Config.ProtocolVersionsAddr),
+		Guardian:              guardian,
+	}
+
+	return superchainRoles, nil
+}
+
+func (c *Intent) SetInitValues(l2ChainIds []common.Hash) error {
+	switch c.IntentConfigType {
+	case IntentConfigTypeCustom:
+		return c.setCustomValues(l2ChainIds)
+
+	case IntentConfigTypeStandard:
+		return c.setStandardValues(l2ChainIds)
+
+	case IntentConfigTypeStrict:
+		return c.setStrictValues(l2ChainIds)
+
+	case IntentConfigTypeStrictOverrides:
+		return c.setStrictValues(l2ChainIds)
+
+	case IntentConfigTypeStandardOverrides:
+		return c.setStandardValues(l2ChainIds)
+
+	case IntentConfigTypeTest:
+		return c.setTestValues(l2ChainIds)
+
+	default:
+		return fmt.Errorf("intent config type not supported")
+	}
+}
+
+// Sets all Intent fields to their zero value with the expectation that the
+// user will populate the values before running 'apply'
+func (c *Intent) setCustomValues(l2ChainIds []common.Hash) error {
+	c.L1ContractsLocator = &artifacts.Locator{Tag: "undefined"}
```

Comment:
What's going on here? This `undefined` construct seems like a smell to me since it leaks implementation details of `Intent` construction.

Reply:
This custom `MarshalText` method fails right now if I don't set the `Tag` string to something:
```
func (a *Locator) MarshalText() ([]byte, error) {
	if a.URL != nil {
		return []byte(a.URL.String()), nil
	}

	if a.Tag != "" {
		return []byte("tag://" + a.Tag), nil
	}

	return nil, fmt.Errorf("no URL, path or tag set")
}
```
I want the field to be written to the toml file so that it nudges the user to fill it out. What is a good placeholder value to set? Should I just delete the `if a.Tag != nil` conditional check?

Reply:
Can you also help me understand the reasoning behind the logic [here](https://github.com/ethereum-optimism/optimism/blob/develop/op-deployer/pkg/deployer/state/intent.go#L78-L92)? I think its relevant to this thread and some of it could potentially could be moved into `intent.validateStandardValues` method

Reply:
Pushed a refactor commit related to this code: https://github.com/ethereum-optimism/optimism/pull/12970/commits/a74e72cd44dd00fcb35b0c29df3c69fcdab858e5